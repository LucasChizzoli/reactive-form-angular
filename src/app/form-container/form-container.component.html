<div class="form-container-wrapper">
  <h3 class="form-container-title">{{formTitle}}</h3>
  <ng-content></ng-content>
</div>
