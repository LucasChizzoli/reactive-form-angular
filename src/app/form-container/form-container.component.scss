.form-container-wrapper {
  min-width: 400px;

  .form-container-title {
    font-size: 2rem;
    text-align: center;
    margin: 0;
  }
}
