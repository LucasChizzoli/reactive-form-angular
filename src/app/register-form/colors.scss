$primary: #478C5C;
$primary-dark: #013A20;
$secondary-dark: #BACC81;
$secondary: #CDD193;
$grey-dark: #cccccc;
$grey-light: #f2f2f2;
$white: #ffffff;
$warn: #de5353;
