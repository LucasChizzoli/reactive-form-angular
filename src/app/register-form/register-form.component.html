<app-form-container formTitle="REGISTER">
  <form [formGroup]="form">
    <div *ngFor="let control of fields">
      <label>
        <span class="form-label">{{control.display || control.name | titlecase}}</span>
        <input [type]="getType(control)" class="form-input" [formControlName]="control.name">
      </label>
      <div
        *ngIf="form.get(control.name)?.invalid &&
        (form.get(control.name)?.dirty || form.get(control.name)?.touched)">
        <p class="form-error" *ngIf="form.get(control.name)?.errors">
          {{getErrorMessage(control, form)}}
        </p>
      </div>
    </div>
    <div class="actions-wrapper">
      <button
        class="register-button"
        type="button"
        (click)="submit()">
        REGISTER
      </button>
    </div>
  </form>
  <div class="register-footer">
    <span>Already have an account? <a href="#" class="form-link">Login now</a></span>
  </div>
</app-form-container>
