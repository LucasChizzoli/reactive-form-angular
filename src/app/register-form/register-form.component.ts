import { Component } from '@angular/core';
import { Validators } from '@angular/forms';
import { FormValidations } from '../utils/form-validations';
import { FormManagerComponent } from '../form-manager.component';

@Component({
  selector: 'app-register-form',
  templateUrl: './register-form.component.html',
  styleUrls: ['./register-form.component.scss']
})
export class RegisterFormComponent extends FormManagerComponent {
  /* ------ INSTRUCTIONS ------
    Add, edit or remove the items in the FIELDS array to configure the form
    Name = Unique identifier. Spaces are not allowed.
    Display (Optional) = This is used for the label of the form. If not present Name will be used instead.
    Validators (Optional) = Validators for the input. Available built-in validators https://angular.io/api/forms/Validators
    You can also create your own like passwordConfirmValidation
  */
  fields = [
    {
      name: 'email',
      validators: [Validators.email, Validators.required]
    },
    {
      name: 'password',
      validators: [Validators.required, Validators.minLength(8)]
    },
    {
      display: 'Confirm Password',
      name: 'confirmPassword',
      validators: [
        Validators.required,
        Validators.minLength(8),
        FormValidations.passwordConfirmValidation(this.form),
      ]
    },
  ];

  constructor() {
    super();
    /* DO NOT REMOVE THIS CODE
      Call init form to initialize and render the form
    */
    this.initForm(this.fields);
  }
}
