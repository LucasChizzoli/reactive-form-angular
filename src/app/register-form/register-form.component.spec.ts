import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RegisterFormComponent } from './register-form.component';
import { FormContainerComponent } from '../form-container/form-container.component';
import { FormManagerComponent } from '../form-manager.component';
import { FormsModule, ReactiveFormsModule, Validators } from '@angular/forms';

describe('RegisterFormComponent', () => {
  let component: RegisterFormComponent;
  let fixture: ComponentFixture<RegisterFormComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        FormsModule,
        ReactiveFormsModule
      ],
      declarations: [ RegisterFormComponent, FormContainerComponent, FormManagerComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(RegisterFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should have email and password and confirm password', () => {
    component.fields = [
      {
        name: 'email',
        validators: [Validators.email, Validators.required]
      },
      {
        name: 'password',
        validators: [Validators.required, Validators.minLength(8)]
      },
      {
        name: 'confirmPassword',
        display: 'Confirm Password',
        validators: [Validators.required, Validators.minLength(8)]
      },
    ];
    fixture.detectChanges();
    expect(component).toMatchSnapshot();
  });
});
