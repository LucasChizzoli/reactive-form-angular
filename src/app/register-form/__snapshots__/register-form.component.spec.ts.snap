// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterFormComponent should have email and password and confirm password 1`] = `
RegisterFormComponent {
  "FIELDS": Array [
    Object {
      "name": "email",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
    Object {
      "name": "password",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
    Object {
      "display": "Confirm Password",
      "name": "confirmPassword",
      "validators": Array [
        [Function],
        [Function],
        [Function],
      ],
    },
  ],
  "fields": Array [
    Object {
      "name": "email",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
    Object {
      "name": "password",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
    Object {
      "display": "Confirm Password",
      "name": "confirmPassword",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
  ],
  "form": FormGroup {
    "_composedAsyncValidatorFn": null,
    "_composedValidatorFn": null,
    "_hasOwnPendingAsyncValidator": false,
    "_onCollectionChange": [Function],
    "_onDisabledChange": Array [],
    "_parent": null,
    "_rawAsyncValidators": null,
    "_rawValidators": null,
    "controls": Object {
      "confirmPassword": FormControl {
        "_composedAsyncValidatorFn": null,
        "_composedValidatorFn": [Function],
        "_hasOwnPendingAsyncValidator": false,
        "_onChange": Array [
          [Function],
          [Function],
        ],
        "_onCollectionChange": [Function],
        "_onDisabledChange": Array [
          [Function],
          [Function],
        ],
        "_parent": [Circular],
        "_pendingValue": "",
        "_rawAsyncValidators": null,
        "_rawValidators": Array [
          [Function],
          [Function],
          [Function],
        ],
        "errors": Object {
          "required": true,
        },
        "pristine": true,
        "status": "INVALID",
        "statusChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "touched": false,
        "value": "",
        "valueChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
      "email": FormControl {
        "_composedAsyncValidatorFn": null,
        "_composedValidatorFn": [Function],
        "_hasOwnPendingAsyncValidator": false,
        "_onChange": Array [
          [Function],
          [Function],
        ],
        "_onCollectionChange": [Function],
        "_onDisabledChange": Array [
          [Function],
          [Function],
        ],
        "_parent": [Circular],
        "_pendingValue": "",
        "_rawAsyncValidators": null,
        "_rawValidators": Array [
          [Function],
          [Function],
        ],
        "errors": Object {
          "required": true,
        },
        "pristine": true,
        "status": "INVALID",
        "statusChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "touched": false,
        "value": "",
        "valueChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
      "password": FormControl {
        "_composedAsyncValidatorFn": null,
        "_composedValidatorFn": [Function],
        "_hasOwnPendingAsyncValidator": false,
        "_onChange": Array [
          [Function],
          [Function],
        ],
        "_onCollectionChange": [Function],
        "_onDisabledChange": Array [
          [Function],
          [Function],
        ],
        "_parent": [Circular],
        "_pendingValue": "",
        "_rawAsyncValidators": null,
        "_rawValidators": Array [
          [Function],
          [Function],
        ],
        "errors": Object {
          "required": true,
        },
        "pristine": true,
        "status": "INVALID",
        "statusChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "touched": false,
        "value": "",
        "valueChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
    },
    "errors": null,
    "pristine": true,
    "status": "INVALID",
    "statusChanges": EventEmitter_ {
      "__isAsync": false,
      "_isScalar": false,
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "touched": false,
    "value": Object {
      "confirmPassword": "",
      "email": "",
      "password": "",
    },
    "valueChanges": EventEmitter_ {
      "__isAsync": false,
      "_isScalar": false,
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
  },
  "getErrorMessage": [Function],
  "getType": [Function],
  "submitted": EventEmitter_ {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
}
`;
